### ISSUE TITLE
Web Support V1

### ISSUE THREAD

**Issue by EvanBacon** [expo/expo#6782]
[Web support beta](https://blog.expo.io/expo-cli-and-sdk-web-support-beta-d0c588221375) has been great, but it's time to make it production ready! Here I've outlined the steps required to creating a production ready system for building basic web apps with Expo (a stable MVP). There are also a few completed V2 features which we can move to another place after V1 is complete.

## Navigation

This has elevated to its own category.

- Examples
  - [x] Next.js for web & React Navigation for native
  - [x] React Router for web & React Navigation for native
  - [x] Drawer Navigator example
  - [x] Using Media Queries to change navigation
- React Navigation
  - [x] History support basic navigators
  - [x] History support for nested navigators

## PWA

- [x] Upgrade to `webpack@4.42.0`
- [x] CLI for creating assets. https://github.com/expo/expo-cli/pull/1686
- [x] Improved PWA generation in Webpack. https://github.com/expo/expo-cli/pull/1686

## Documentation

- [x] Adding web support
- [x] Starting a web project
- [x] Building for web
- [x] [Publishing to web hosting platforms](https://docs.expo.io/versions/v36.0.0/guides/deploying-websites/)
- [x] [Customizing Webpack](https://docs.expo.io/versions/latest/guides/customizing-webpack/)
- [x] [Babel Preset Documentation](https://www.npmjs.com/package/babel-preset-expo)
- Styling options: 
  - [x] `styled-components` vs React Native Style Sheet
  - [x] [Using hooks to create pseudo class effects](https://blog.expo.io/css-pseudo-class-effects-in-expo-for-web-56649f88eb6b) like hover, focus, and active states. 
- [x] Routing and navigation
  - [x] [Example of using React Router](https://github.com/expo/examples/pull/42)
- [x] [Performance tips](https://docs.expo.io/versions/latest/guides/web-performance/)
- [x] [Usage in the Bare-workflow](https://docs.expo.io/versions/v36.0.0/bare/using-web/)
- [x] [Using TypeScript](https://docs.expo.io/versions/v36.0.0/guides/typescript/)
- [x] [Video tutorial](https://www.youtube.com/watch?v=Czih6w57P9A)
- [ ] PWAs

## Testing

- [x] [unit] [Web support in `jest-expo`](https://blog.expo.io/testing-universal-react-native-apps-with-jest-and-expo-113b4bf9cc44)
- [x] [render] [Support Enzyme](https://www.npmjs.com/package/jest-expo-enzyme) `jest-expo-enzyme`
- [x] [E2E] [Support Puppeteer](https://www.npmjs.com/package/jest-expo-puppeteer) `jest-expo-puppeteer`
- [x] [Storybook example](https://github.com/expo/examples/tree/master/with-storybook)

## SDK

- [x] Add unavailability errors to any native functionality that doesn't exist on web
- [x] Support all **basic** primitive features in the Expo ecosystem
- Improve React Native for web
  - [x] Fix tree-shaking

### V2

- Improve React Native for web
  - [ ] Slider component
  - [ ] Momentum scroll events
  - [ ] Fixing tint Color bug on safari
- Support extended ecosystem 
  - [x] [`react-native-gesture-handler`](https://github.com/software-mansion/react-native-gesture-handler/pulls?q=is%3Apr+label%3A%22Platform%3A+web%22+is%3Aclosed)
  - [ ] [`react-native-reanimated`](https://github.com/software-mansion/react-native-reanimated/pulls?utf8=%E2%9C%93&q=is%3Apr+%5Bweb%5D+is%3Aclosed+)
  - [x] [`react-native-safe-area-context`](https://github.com/th3rdwave/react-native-safe-area-context/pull/2)
  - [x] [`react-native-web-hooks`](https://github.com/EvanBacon/react-native-web-hooks/)
  - [x] `react-native-shared-element`
  - [x] [`react-native-svg`](https://github.com/react-native-community/react-native-svg/pull/950)
  - [ ] `@react-native-community/web-view`
  - [ ] React Navigation V5
- [ ] Improve/Add experimental packages
  - [ ] bar code scanner
  - [ ] face detector 
- Support third-party SDK modules
  - [ ] expo-facebook
  - [ ] expo-google
  - [ ] Sentry
  - [ ] Amplitude
  - [ ] expo-apple-authentication
    - [ ] Merge `react-apple-authentication` into `expo-apple-authentication`
  - [ ] Branch
  - [ ] Facebook Ads
  - [ ] Push Notifications
  - [ ] Payments - Stripe - expo-payments-stripe

## Custom Workflows

- [x] Support basic Gatsby 
  - [x] Document getting started
  - [x] Using Expo SDK
  - [x] Building for deployment
  - [x] [Example](https://github.com/expo/examples/tree/master/with-gatsby)
- [x] Support basic Next.js
  - [x] Document getting started
  - [x] Using Expo SDK
  - [x] Building for deployment
  - [x] [Example](https://github.com/expo/examples/tree/master/with-nextjs)
- [x] [Support basic Preact](https://github.com/expo/expo/blob/master/docs/pages/versions/unversioned/guides/using-preact.md)

### V2

- Support basic Electron
  - [x] [Document getting started](https://dev.to/evanbacon/making-desktop-apps-with-electron-react-native-and-expo-5e36)
  - [x] Using Expo SDK
  - [ ] Building for deployment
  - [x] [Example](https://github.com/expo/examples/tree/master/with-electron)
- [x] [Support Preact](https://github.com/expo/expo/pull/6718)

## Tooling

- [x] Starting a project with `expo start:web` & `expo start`
  - [x] E2E test
- [x] Building with `expo build:web`
  - [x] E2E test
  - [x] Generate PWA
  - [x] Generate favicon **V2**
- [x] [Custom Webpack config](https://github.com/expo/expo-cli/tree/master/packages/webpack-config)
  - [x] mixable functions for adding Expo support to other Webpack configs
  - [x] Unit test shape
- `app.json` web functionality
  - [x] Move all webpack config options out https://github.com/expo/expo-cli/pull/1510
  - [x] Remove all dangerous values

### Snack

- [x] Add web support

### Client

- [x] Show running expo web processes and open to web browser

**Comment by alejomendoza** [comment]
Awesome video and updates! I couldn't use the `Video` component in web, couldn't tell if it was supported in SDK 36, docs say it is. Great work :)

**Comment by AltairTodescatto** [comment]
Very cool, cant wait to get more of this tool.

**Comment by arlyon** [comment]
How likely is it that we could see "proper" (as opposed to react native) storybook support via expo? Surely the web target would make that possible?

**Comment by tehOPEologist** [comment]
@EvanBacon are there any plans to integrate Fast Refresh for web?  i've seen some webpack plugins that give experimental support, but haven't had luck getting all the kinks worked out.  excited to see you @ RNA in march!  ^_^

**Comment by EvanBacon** [comment]
@tehOPEologist https://github.com/expo/expo-cli/issues/1498

**Comment by liming-developer** [comment]
Syntax Error: Adjacent JSX elements must be wrapped in an enclosing tag. Did you want a JSX fragment <>...</>?
How can i solve it?

**Comment by AltairTodescatto** [comment]
@liming-developer You had to wrap all the elements rendered in a page into a single enclosing tag. Just put a <> ... </> around your code into the render. A <div> ... </div> works just as fine

**Comment by ningacoding** [comment]
support for expo install @react-native-community/viewpager ??? on Web

**Comment by abduraufsherkulov** [comment]
`ImagePicker` doesn't work as expected. Even if I disable option `base64`, it still provides me `base64` data instead of `files object`.

**Comment by elgohary99** [comment]
if i using web browser to run..that way same android studio to run

**Comment by raarts** [comment]
What are V1 and V2, do those relate to the Web support? Will there be a Web Support V2?

**Comment by nonotest** [comment]
I have created a small app with rnw, nextjs and navigation v5, it is very impressive.
Needs a little configuration but still. very good job!

Any chance we could add better support for react-native-maps in the ecosystem category?
It looks like the project needs some workaround with proptypes and requireNativeComponent usage. I couldn't find any web specific issue on their repo.

I haven't really been able to make it work so far but happy to help if needed

**Comment by EvanBacon** [comment]
@nonotest, the expo fork of react-native-maps has partial web support for Leaflet. It's a lot of work to add providers so I am mostly prioritizing packages in the Expo ecosystem that I have more freedom to edit.

@raarts V1 is what we're working towards right now, V2 means it's on the roadmap but not a priority at the moment.

**Comment by RWOverdijk** [comment]
Yeeeees please. Does this include easier setup for useLinking? Because right now it's near impossible to set it up properly with nested routers and a correct history.

If so, I'll repeat my Yeeeees please 😄

**Comment by Albert-Gao** [comment]
Awesome support, everything works so well so far.

One question, any recommended lib for storage? The doc said AsyncStorage is deprecated. But I have a problem of using `react-native-community/async-storage` v2, got an `AsyncStorage is null` error in a managed workflow. Maybe due to the native link, but I don't want to eject for this.

I currently just use AsyncStorage for native, and LocalStorage for web. I have very tiny info to save, works great so far.

1. Any recommend lib for this which works with managed workflow?
2. What does the deprecate mean here? The end goal is to remove this API in the end? How long would it be?

Thanks :)

**Comment by Mukulagarwal1980** [comment]
> I have created a small app with rnw, nextjs and navigation v5, it is very impressive.
> Needs a little configuration but still. very good job!
> 
> Any chance we could add better support for react-native-maps in the ecosystem category?
> It looks like the project needs some workaround with proptypes and requireNativeComponent usage. I couldn't find any web specific issue on their repo.
> 
> I haven't really been able to make it work so far but happy to help if needed

Hello, 
Have you implemented custom view for drawer navigation v5

**Comment by Mukulagarwal1980** [comment]
Hello,

I m getting below error when allow for permissions for push notification on web app.

**You must provide `owner` and `slug` in `app.json` to use push notifications on web. Learn more: https://docs.expo.io/versions/latest/guides/using-vapid/.**

Please provide solution if any body have
Thanks in advance

**Comment by Sunil-prajapati** [comment]
Hello to all,
I am facing below given problem someone suggests me the solution 

37.0.0 is not a valid sdk version option are 35.0.0

**Comment by mufeez-amjad** [comment]
I keep getting when running `expo start:web`:
```
Module not found: Can't resolve 'react-native' in 'app/node_modules/react-native-svg/lib/module'
```

**Comment by pyros2097** [comment]
@mufeez-amjad You might need this `react-native-svg-transformer`

**Comment by pyros2097** [comment]
I've got history support for basic navigators done and working. I was using react-navigation@v4 as a reference. It allows to reload the url with state preserved as along as the screen params are simple objects.
```js
import React, { useState, useRef, useEffect } from 'react';
import { Platform } from 'react-native';
import { NavigationContainer } from '@react-navigation/native';
import { createBrowserHistory, createMemoryHistory } from 'history';
import * as queryString from 'query-string';
const history = Platform.OS === 'web' ? createBrowserHistory() : createMemoryHistory();

export const getActiveScreen = (state) => {
  const route = state.routes[state.index];
  if (route.state) {
    return getActiveScreen(route.state);
  }
  return { name: route.name, params: route.params };
};

export const getInitialScreen = (initial) => {
  if (Platform.OS === 'web') {
    const path = window.location.pathname.replace('/', '');
    const params = queryString.parse(window.location.search);
    return { name: path ? path : initial, params };
  }
  return { name: initial, params: {} };
};

export const getUrl = (name, params) => {
  const query = queryString.stringify(params);
  return name[0].toLowerCase() + name.slice(1) + (query ? '?' + query : '');
};

export default WebNavigationContainer = ({ setNavigation, children }) => {
  const [currentScreen, setCurrentScreen] = useState(null);
  const navigationRef = useRef();
  useEffect(() => {
    if (navigationRef.current) {
      const state = navigationRef.current.getRootState();
      setCurrentScreen(getActiveScreen(state));
      setNavigation(navigationRef.current);
    }
    return history.listen((location, action) => {
      if (action === 'POP') {
        navigationRef.current.goBack();
      }
    });
  }, [navigationRef !== undefined && navigationRef !== null]);
  const onStateChange = (state) => {
    const nextScreen = getActiveScreen(state);
    if (currentScreen.name !== nextScreen.name) {
      if (!(nextScreen.name.indexOf('Navigator') > -1)) {
        let navigate = history.push;
        // if (currentScreen && currentScreen.name.indexOf('Navigator') > -1) {
        //   navigate = history.replace;
        // }
        const url = getUrl(nextScreen.name, nextScreen.params);
        navigate(url);
        setCurrentScreen(nextScreen);
      }
    }
  };
  return (
    <NavigationContainer ref={navigationRef} onStateChange={onStateChange}>
      {children}
    </NavigationContainer>
  );
};

export const WebNavigator = ({ Comp, screenOptions, initialRouteName, initialRouteParams, children, ...props }) => {
  return (
    <Comp screenOptions={screenOptions} initialRouteName={initialRouteName} {...props}>
      {React.Children.map(children, (child) => {
        const initialParams = initialRouteName === child.props.name ? initialRouteParams : {};
        return React.cloneElement(child, { initialParams });
      })}
    </Comp>
  );
};
```

Example:
```js
import React, { useState } from 'react';
import { View, Text, Button } from 'react-native';
import { createStackNavigator } from '@react-navigation/stack';
import { WebNavigationContainer, WebNavigator, getInitialScreen } from './navigation';

const Home = ({ navigation }) => (
  <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
    <Text>Home</Text>
    <Button title="Login" onPress={() => navigation.navigate('Login', { email: 'abc' })} />
  </View>
);
const Login = ({ route: { params } }) => (
  <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
    <Text>Login</Text>
    <Text>Email: {params.email}</Text>
  </View>
);

const Stack = createStackNavigator();

export default function App() {
  const [navigation, setNavigation] = useState(null);
  const { name, params } = getInitialScreen();
  return (
    <WebNavigationContainer setNavigation={setNavigation}>
      <WebNavigator Comp={Stack.Navigator} initialRouteName={name} initialRouteParams={params}>
        <Stack.Screen name="Home" component={Home} />
        <Stack.Screen name="Login" component={Login} />
      </WebNavigator>
    </WebNavigationContainer>
  );
}
```

**Comment by hyochan** [comment]
<img width="482" alt="Screen Shot 2020-05-16 at 10 16 04 PM" src="https://user-images.githubusercontent.com/27461460/82120685-ed5d8480-97c2-11ea-80d1-1fc72f3dea86.png">

Is `react-native-svg-transformer` not working with `expo-web`? 
I am facing the above error and trying to add below [code](https://github.com/kristerkari/react-native-svg-example/blob/c97ab040f0fb7f0438991ef55344f68e846a8e88/webpack.config.js#L28) by [customizing webpack](https://docs.expo.io/guides/customizing-webpack).

**Comment by hyochan** [comment]
Ok customizing webpack as above didn't help. I've found the [issue](https://github.com/kristerkari/react-native-svg-transformer/issues/70) in `react-native-svg-transformmer`.

**Comment by pyros2097** [comment]
@hyochan It works for me. I took this as an example https://github.com/kristerkari/react-native-svg-example. I had initial hiccups but it worked later on. This is my config.
```
const { getDefaultConfig } = require('metro-config');

module.exports = (async () => {
  const {
    resolver: { sourceExts, assetExts }
  } = await getDefaultConfig();
  return {
    transformer: {
      getTransformOptions: async () => ({
        transform: {
          experimentalImportSupport: false,
          inlineRequires: false
        }
      }),
      babelTransformerPath: require.resolve('react-native-svg-transformer')
    },
    resolver: {
      assetExts: assetExts.filter(ext => ext !== 'svg'),
      sourceExts: [...sourceExts, 'svg']
    }
  };
})();
```

**Comment by hyochan** [comment]
@pyros2097 Wow! Thanks for sharing so quickly. I've also found the issue as I've posted [above](https://github.com/expo/expo/issues/6782#issuecomment-629646298).

By the way, I am facing another problem in `react-native-svg`.
<img width="215" alt="Screen Shot 2020-05-16 at 11 40 46 PM" src="https://user-images.githubusercontent.com/27461460/82122472-db81de80-97ce-11ea-9425-894a30156529.png">
<img width="69" alt="Screen Shot 2020-05-16 at 11 41 02 PM" src="https://user-images.githubusercontent.com/27461460/82122473-dd4ba200-97ce-11ea-8633-966f53a032b3.png">

When I set the size manually, as shown above, the image is cutter instead of resizing.
<img width="82" alt="Screen Shot 2020-05-16 at 11 42 35 PM" src="https://user-images.githubusercontent.com/27461460/82122488-fa807080-97ce-11ea-89ad-a990b6d47cd9.png">

**Comment by pyros2097** [comment]
@hyochan Your Welcome.
Hmmm....I have come across this and I do remember it was something to do with the SVG rather than the width and height prop.  But I think you need to make sure your SVG doesn't have width and height set and also x and y a are 0. 
Also try tweaking the viewBox on the svg it seems to me that is the culprit. But this shouldn't happen.
I use it like this and it works fine. In my case I needed the height shorter so I changed the view box. But in itself the DVG should scale properly without cutting off just like those icons we use.
```js
<LogoCircle
            viewBox={Platform.OS === 'web' ? '0 0 1000 1000' : '0 0 1000 630'}
            width="100%"
            height={Platform.OS === 'web' ? '550' : height}
            style={{ marginTop: marginTop }}
          />
```

**Comment by hyochan** [comment]
@pyros2097 Well
```
<svg width="52" height="52" viewBox="0 0 52 52" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M48.4453 15.8438H46.1094C39.5555 13.1483 32.7915 11.7812 26 11.7812C19.2085 11.7812 12.4485 13.1483 5.89062 15.8438H3.55469C2.61193 15.8438 1.70778 16.2183 1.04114 16.8849C0.374511 17.5515 0 18.4557 0 19.3984L0 32.6016C0 33.5443 0.374511 34.4485 1.04114 35.1151C1.70778 35.7817 2.61193 36.1562 3.55469 36.1562H5.89062C12.4445 38.8517 19.2085 40.2188 26 40.2188C32.7915 40.2188 39.5515 38.8517 46.1094 36.1562H48.4453C49.3881 36.1563 50.2922 35.7817 50.9589 35.1151C51.6255 34.4485 52 33.5443 52 32.6016V19.3984C52 18.4557 51.6255 17.5515 50.9589 16.8849C50.2922 16.2183 49.3881 15.8438 48.4453 15.8438V15.8438ZM44.8906 24.9844H7.10938V21.5424C19.5331 18.6161 32.4669 18.6161 44.8906 21.5424V24.9844ZM7.10938 27.0156H44.8906V30.4576C38.6997 31.9165 32.3605 32.653 26 32.6523C19.6395 32.653 13.3003 31.9165 7.10938 30.4576V27.0156ZM44.8906 17.5419V19.4553C32.4583 16.6034 19.5417 16.6034 7.10938 19.4553V17.5419C13.2793 15.0668 19.631 13.8125 26 13.8125C32.369 13.8125 38.7207 15.0668 44.8906 17.5419ZM2.03125 32.6016V19.3984C2.03125 18.9944 2.19175 18.6069 2.47745 18.3212C2.76315 18.0355 3.15065 17.875 3.55469 17.875H5.07812V34.125H3.55469C3.15065 34.125 2.76315 33.9645 2.47745 33.6788C2.19175 33.3931 2.03125 33.0056 2.03125 32.6016ZM7.10938 34.4581V32.5447C13.3056 33.9668 19.6427 34.6843 26 34.6836C32.3573 34.6843 38.6944 33.9668 44.8906 32.5447V34.4581C38.7207 36.9332 32.369 38.1875 26 38.1875C19.631 38.1875 13.2793 36.9332 7.10938 34.4581ZM49.9688 32.6016C49.9687 33.0056 49.8082 33.3931 49.5225 33.6788C49.2368 33.9645 48.8494 34.125 48.4453 34.125H46.9219V17.875H48.4453C48.8494 17.875 49.2368 18.0355 49.5225 18.3212C49.8082 18.6069 49.9688 18.9944 49.9688 19.3984V32.6016Z" fill="#000"/>
</svg>
```
Changing above svg to below also didn't help.
```
<svg viewBox="0 0 0 0" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M48.4453 15.8438H46.1094C39.5555 13.1483 32.7915 11.7812 26 11.7812C19.2085 11.7812 12.4485 13.1483 5.89062 15.8438H3.55469C2.61193 15.8438 1.70778 16.2183 1.04114 16.8849C0.374511 17.5515 0 18.4557 0 19.3984L0 32.6016C0 33.5443 0.374511 34.4485 1.04114 35.1151C1.70778 35.7817 2.61193 36.1562 3.55469 36.1562H5.89062C12.4445 38.8517 19.2085 40.2188 26 40.2188C32.7915 40.2188 39.5515 38.8517 46.1094 36.1562H48.4453C49.3881 36.1563 50.2922 35.7817 50.9589 35.1151C51.6255 34.4485 52 33.5443 52 32.6016V19.3984C52 18.4557 51.6255 17.5515 50.9589 16.8849C50.2922 16.2183 49.3881 15.8438 48.4453 15.8438V15.8438ZM44.8906 24.9844H7.10938V21.5424C19.5331 18.6161 32.4669 18.6161 44.8906 21.5424V24.9844ZM7.10938 27.0156H44.8906V30.4576C38.6997 31.9165 32.3605 32.653 26 32.6523C19.6395 32.653 13.3003 31.9165 7.10938 30.4576V27.0156ZM44.8906 17.5419V19.4553C32.4583 16.6034 19.5417 16.6034 7.10938 19.4553V17.5419C13.2793 15.0668 19.631 13.8125 26 13.8125C32.369 13.8125 38.7207 15.0668 44.8906 17.5419ZM2.03125 32.6016V19.3984C2.03125 18.9944 2.19175 18.6069 2.47745 18.3212C2.76315 18.0355 3.15065 17.875 3.55469 17.875H5.07812V34.125H3.55469C3.15065 34.125 2.76315 33.9645 2.47745 33.6788C2.19175 33.3931 2.03125 33.0056 2.03125 32.6016ZM7.10938 34.4581V32.5447C13.3056 33.9668 19.6427 34.6843 26 34.6836C32.3573 34.6843 38.6944 33.9668 44.8906 32.5447V34.4581C38.7207 36.9332 32.369 38.1875 26 38.1875C19.631 38.1875 13.2793 36.9332 7.10938 34.4581ZM49.9688 32.6016C49.9687 33.0056 49.8082 33.3931 49.5225 33.6788C49.2368 33.9645 48.8494 34.125 48.4453 34.125H46.9219V17.875H48.4453C48.8494 17.875 49.2368 18.0355 49.5225 18.3212C49.8082 18.6069 49.9688 18.9944 49.9688 19.3984V32.6016Z" fill="#000"/>
</svg>
```

**Comment by hyochan** [comment]
@pyros2097 Removing `width` and `height` in svg did rock! Thanks 🥇 . However, it wasn't necessary in `ios` and `android`.

**Comment by hyochan** [comment]
I just found out that `Alert.alert` isn't working as in `ios` and `android` in `web`. I just do nothing.

Today, I am confused if this kind of issue should be posted in `react-native-web` or in `expo`. @EvanBacon As an expo-web lead, where would you want me to write issues?

**Comment by inglesuniversal** [comment]
@EvanBacon first of all, I was very excited watching your conference in January of this year in Berlin, where you showcased all the latest features of Expo that this year have gone into another dimension of progress. Way to go!!!

From your experience: what is the best approach to handle WebView in RN, which works beautifully on iOS and Android... but for the web (since current library does not support web yet).

Is there a way to switch or load iFrame when the App is running on the browser?

Any feedback is greatly appreciated.

STAY SAFE!!! The world needs Developers like you!

Hugo Barbosa

**Comment by Albert-Gao** [comment]
Just a quick question, how to update the HTML template, I need to add a <div> there. Thanks

**Comment by arbre-abuche** [comment]
Expo web is in beta, please report any bugs or missing features on the Expo repo.       │
You can follow the V1 release for more info: https://github.com/expo/expo/issues/6782  
----------------------------------------------------
One error is not fixed on my computer :
----------------------------------------------------
Emulator: E0612 13:24:33.694699843 25920 socket_utils_common_posix.cc:201] check for SO_REUSEPORT: {"created":"@1591961073.694681972","description":"SO_REUSEPORT unavailable on compiling system","file":"/mnt/tmpfs/src/android/emu-master-dev/external/grpc/src/core/lib/iomgr/socket_utils_common_posix.cc","file_line":169}

**Comment by rbed23** [comment]
Alert object is working in Android emulator but not with Web

**Comment by shubhamdeol** [comment]
How can I ignore unsupported dependencies expo web bare work flow. I am getting ```no loader are configured error``` from one of dependency. due to which I am unable to run my web version. 

```
/react-native-scrollable-tabview/lib/index.js 77:13
Module parse failed: Unexpected token (77:13)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|   }
| 
>   _onRefresh = () => {
|     //if there is not pullToRefresh function do nothing
|     if (!this.props.pullToRefresh)

```

**Comment by Macattack1999-teck** [comment]
Getting this error on web:
index.js:1 TypeError: react_native_web_dist_exports_Animated__WEBPACK_IMPORTED_MODULE_12__.default.subtract is not a function

Pinpoints to this:
const YOLO = Animated.modulo(
      | ^  138 |     Animated.divide(
  139 |       Animated.modulo(Animated.subtract(_angle, angleOffset), oneTurn),
  140 |       new Animated.Value(angleBySegment)

**Comment by WillzMu** [comment]
I am getting an issue where the app open in the web automatically. Is there a way to turn this feature off?

**Comment by raarts** [comment]
@WillzMu You really need to give more information here. What kind of app? Sample code?

**Comment by WillzMu** [comment]
Sorry about that @raarts . I was on an older version of expo-cli. Cant remember exactly what version but I think it was 3.20.*  I ran `expo init <app-name>` to create a project. After that ran `expo start`.  Then I noticed the project was opening automatically in the browser. I ran `npm install -g expo-cli` and the project is not opening up automatically. I noticed a Stack Overflow question about this as well. https://stackoverflow.com/questions/62656214/expo-web-starts-automatically

**Comment by city0666** [comment]
hi expo background location not working

**Comment by aliJawadDev** [comment]
Hello, Qr Is not loading in cmd after using command **expo start**

**Comment by NishargShah** [comment]
```
React does not recognize the `autoCompleteType` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `autocompletetype` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
```

@EvanBacon bug fix :)

**Comment by OrlandoGareca** [comment]
info @expo/ngrok-bin-freebsd-x64@2.2.8: The platform "win32" is incompatible with this module.
info "@expo/ngrok-bin-freebsd-x64@2.2.8" is an optional dependency and failed compatibility check. Excluding it from installation.


problem library ngrok  intall in windows 10

**Comment by brentvatne** [comment]
@OrlandoGareca - no, that's not what is happening. this is just a warning that npm shows for optional packages, you will see it for macos and linux if you're using windows, you can ignore it.

**Comment by eguonzy** [comment]
i opened my app via expo web and some layout attributes changed and it doesnt consume the api

**Comment by renegoretzka** [comment]
Image is not showing up in the web view, because its set to position: absolute; instead of position: relative; as it should be with react-native.

**Comment by G4BB3R** [comment]
react-native-modal-datetime-picker doesn't work for web

**Comment by brentvatne** [comment]
@G4BB3R - that is unrelated to expo, you should post an issue on the react-native-modal-datetime-picker repository. we can't go around and update everybody's repos to support web, that is not feasible

**Comment by mikemathewson** [comment]
DIVs do not have relative position by default, so properties such as top and left on have no effect. Also SafeAreaView doesn't seem to be working properly, the padding is set to a value such as "env(safe-area-inset-top)". I would expect it to just work as a standard View for web.

**Comment by mahermassoud** [comment]
Looks like console.log() doesn't work for me for web but it does for iOS.

Meaning when I do a console.log() message, it shows up on the CLI when i open the app on my phone through tunnel, but it does not show up when I open as a web app (local or not). I'm on Mac OS 11.0.1 and using expo version 4.2.1

thanks

**Comment by brentvatne** [comment]
@mahermassoud - console.log on web goes to your browser developer tools logs. we show logs in expo-cli for ios and android because there is no equivalent to this in those cases. it's much better to use your developer tools logs than having to serialize and print the logs in expo-cli terminal. in other words, this is intentional, expected behavior

**Comment by mahermassoud** [comment]
@brentvatne OMG Thanks so much!!!!!!!!!!!

**Comment by nikurou** [comment]
Ran into a strange issue on the web version while going thru a Udemy RN course, hopefully others are able to replicate and its not just me. Basically, when you spread your style properties over a view like so...
```
<View style={{ ...styles.container, ...{ backgroundColor: props.color } }}>
    //some content
</View>
```
...generally the second style object you spread over will overwrite any overlapping properties of the first style object, but otherwise keeps all the style properties from the first (styles.container) that weren't overwritten by the second. While this works as intended on my android emulator in the screenshot below, the behavior in web seems to be that it decided to ditch the {styles.container} entirely, and only apply {backgroundColor: props.color}?

For reference the style of styles.container was: 
```
container: {
    flex: 1,
    borderRadius: 10,
    shadowColor: "black",
    shadowOpacity: 0.26,
    shadowOffset: { width: 0, height: 2 },
    shadowRadius: 10,
    elevation: 3,
    padding: 10,
    justifyContent: "flex-end",
    alignItems: "flex-end",
    backgroundColor: "pink",
  },
```
Image: 
![image_2021-03-08_202414](https://user-images.githubusercontent.com/47164155/110418770-f4ed6300-804c-11eb-8b96-253df5f8e82e.png)

**Comment by ianmartorell** [comment]
@nikurou I'm not sure why that works on native, but it's probably due to the fact that styles aren't actually objects, but optimized references, so you can't spread them. It will work if you use an array instead:

```jsx
<View style={[styles.container, { backgroundColor: props.color }]}>
    //some content
</View>
```

**Comment by vickneswarixbi4** [comment]
Hi, expo go app doesn't get reloading once a plugin is imported

**Comment by alaminsheikh01** [comment]
Hi, when I try to run my app as web then show me error like this: 
```Failed to compile.
C:/Users/alami/OneDrive/Desktop/React native/starter-react-native/node_modules/react-native-gesture-handler/Swipeable.js
Module not found: Can't resolve 'react-native-web/dist/exports/Animated' in 'C:\Users\alami\OneDrive\Desktop\React native\starter-react-native\node_modules\react-native-gesture-handler'```

Any idea please

**Comment by chackerian** [comment]
I'm getting a `Failed to compile. Invalid file signature` error. When any part of the page is changed and it re renders, it loads properly. I'm not sure what is causing this issue.

**Comment by raulvictorrosa** [comment]
When using theme props of styled-components is not working on web.

Example bellow is returning error.
```
export const Title = styled.Text`
  color: ${({theme}) => theme.colors.secondary};
  font-family: ${({theme}) => theme.fonts.boldTitle};
  font-size: ${RFValue(32)}px;
  margin-bottom: 28px;
  line-height: 32px;
`;
```

Is returning the following error.
`TypeError: Cannot read property 'secondary' of undefined`
`TypeError: Cannot read property 'boldTitle' of undefined`

**Comment by mariomurrent-softwaresolutions** [comment]
I'm getting the following error, when I try to start my app with expo web:

```
Failed to compile.
/Users/.../node_modules/react-native/Libraries/Components/TextInput/TextInputState.js
Module not found: Can't resolve '../../Utilities/Platform' in '/Users/.../node_modules/react-native/Libraries/Components/TextInput'
```

Any idea how to solve this?

**Comment by GustavoEklund** [comment]
I'm getting the following error, when I try to start my app with expo web:

`/node_modules/react-native/Libraries/LogBox/UI/LogBoxNotification.js
Module not found: Can't resolve '../../Image/Image' in '/home/gustavoeklund/www/ltw-bank-mobile/node_modules/react-native/Libraries/LogBox/UI'`

**Comment by prithvi-jpg** [comment]
> [Web support beta](https://blog.expo.io/expo-cli-and-sdk-web-support-beta-d0c588221375) has been great, but it's time to make it production ready! Here I've outlined the steps required to creating a production ready system for building basic web apps with Expo (a stable MVP). There are also a few completed V2 features which we can move to another place after V1 is complete.
> 
> ## Navigation
> This has elevated to its own category.
> 
> * Examples
>   
>   * [ ]  Next.js for web & React Navigation for native
>   * [ ]  React Router for web & React Navigation for native
>   * [ ]  Drawer Navigator example
>   * [ ]  Using Media Queries to change navigation
> * React Navigation
>   
>   * [ ]  History support basic navigators
>   * [ ]  History support for nested navigators
> 
> ## PWA
> * [ ]  Upgrade to `webpack@4.42.0`
> * [x]  CLI for creating assets. [Redesign Expo web PWA system expo-cli#1686](https://github.com/expo/expo-cli/pull/1686)
> * [x]  Improved PWA generation in Webpack. [Redesign Expo web PWA system expo-cli#1686](https://github.com/expo/expo-cli/pull/1686)
> 
> ## Documentation
> * [x]  Adding web support
> * [x]  Starting a web project
> * [x]  Building for web
> * [x]  [Publishing to web hosting platforms](https://docs.expo.io/versions/v36.0.0/guides/deploying-websites/)
> * [x]  [Customizing Webpack](https://docs.expo.io/versions/latest/guides/customizing-webpack/)
> * [x]  [Babel Preset Documentation](https://www.npmjs.com/package/babel-preset-expo)
> * Styling options:
>   
>   * [x]  `styled-components` vs React Native Style Sheet
>   * [x]  [Using hooks to create pseudo class effects](https://blog.expo.io/css-pseudo-class-effects-in-expo-for-web-56649f88eb6b) like hover, focus, and active states.
> * [x]  Routing and navigation
>   
>   * [x]  [Example of using React Router](https://github.com/expo/examples/pull/42)
> * [x]  [Performance tips](https://docs.expo.io/versions/latest/guides/web-performance/)
> * [x]  [Usage in the Bare-workflow](https://docs.expo.io/versions/v36.0.0/bare/using-web/)
> * [x]  [Using TypeScript](https://docs.expo.io/versions/v36.0.0/guides/typescript/)
> * [x]  [Video tutorial](https://www.youtube.com/watch?v=Czih6w57P9A)
> * [ ]  PWAs
> 
> ## Testing
> * [x]  [unit] [Web support in `jest-expo`](https://blog.expo.io/testing-universal-react-native-apps-with-jest-and-expo-113b4bf9cc44)
> * [x]  [render] [Support Enzyme](https://www.npmjs.com/package/jest-expo-enzyme) `jest-expo-enzyme`
> * [x]  [E2E] [Support Puppeteer](https://www.npmjs.com/package/jest-expo-puppeteer) `jest-expo-puppeteer`
> * [x]  [Storybook example](https://github.com/expo/examples/tree/master/with-storybook)
> 
> ## SDK
> * [x]  Add unavailability errors to any native functionality that doesn't exist on web
> * [x]  Support all **basic** primitive features in the Expo ecosystem
> * Improve React Native for web
>   
>   * [x]  Fix tree-shaking
> 
> ### V2
> * Improve React Native for web
>   
>   * [ ]  Slider component
>   * [ ]  Momentum scroll events
>   * [ ]  Fixing tint Color bug on safari
> * Support extended ecosystem
>   
>   * [x]  [`react-native-gesture-handler`](https://github.com/software-mansion/react-native-gesture-handler/pulls?q=is%3Apr+label%3A%22Platform%3A+web%22+is%3Aclosed)
>   * [ ]  [`react-native-reanimated`](https://github.com/software-mansion/react-native-reanimated/pulls?utf8=%E2%9C%93&q=is%3Apr+%5Bweb%5D+is%3Aclosed+)
>   * [x]  [`react-native-safe-area-context`](https://github.com/th3rdwave/react-native-safe-area-context/pull/2)
>   * [x]  [`react-native-web-hooks`](https://github.com/EvanBacon/react-native-web-hooks/)
>   * [x]  `react-native-shared-element`
>   * [x]  [`react-native-svg`](https://github.com/react-native-community/react-native-svg/pull/950)
>   * [ ]  `@react-native-community/web-view`
>   * [ ]  React Navigation V5
> * [ ]  Improve/Add experimental packages
>   
>   * [ ]  bar code scanner
>   * [ ]  face detector
> * Support third-party SDK modules
>   
>   * [ ]  expo-facebook
>   * [ ]  expo-google
>   * [ ]  Sentry
>   * [ ]  Amplitude
>   * [ ]  expo-apple-authentication
>     
>     * [ ]  Merge `react-apple-authentication` into `expo-apple-authentication`
>   * [ ]  Branch
>   * [ ]  Facebook Ads
>   * [ ]  Push Notifications
>   * [ ]  Payments - Stripe - expo-payments-stripe
> 
> ## Custom Workflows
> * [x]  Support basic Gatsby
>   
>   * [x]  Document getting started
>   * [x]  Using Expo SDK
>   * [x]  Building for deployment
>   * [x]  [Example](https://github.com/expo/examples/tree/master/with-gatsby)
> * [x]  Support basic Next.js
>   
>   * [x]  Document getting started
>   * [x]  Using Expo SDK
>   * [x]  Building for deployment
>   * [x]  [Example](https://github.com/expo/examples/tree/master/with-nextjs)
> * [x]  [Support basic Preact](https://github.com/expo/expo/blob/master/docs/pages/versions/unversioned/guides/using-preact.md)
> 
> ### V2
> * Support basic Electron
>   
>   * [x]  [Document getting started](https://dev.to/evanbacon/making-desktop-apps-with-electron-react-native-and-expo-5e36)
>   * [x]  Using Expo SDK
>   * [ ]  Building for deployment
>   * [x]  [Example](https://github.com/expo/examples/tree/master/with-electron)
> * [x]  [Support Preact](https://github.com/expo/expo/pull/6718)
> 
> ## Tooling
> * [x]  Starting a project with `expo start:web` & `expo start`
>   
>   * [x]  E2E test
> * [x]  Building with `expo build:web`
>   
>   * [x]  E2E test
>   * [x]  Generate PWA
>   * [ ]  Generate favicon **V2**
> * [x]  [Custom Webpack config](https://github.com/expo/expo-cli/tree/master/packages/webpack-config)
>   
>   * [x]  mixable functions for adding Expo support to other Webpack configs
>   * [x]  Unit test shape
> * `app.json` web functionality
>   
>   * [x]  Move all webpack config options out [Add ability to include modules to the babel loader expo-cli#1510](https://github.com/expo/expo-cli/pull/1510)
>   * [x]  Remove all dangerous values
> 
> ### Snack
> * [x]  Add web support
> 
> ### Client
> * [x]  Show running expo web processes and open to web browser

When is V2 coming cz i want to run my react native app as a pwa on the web, i'm not sure if these issues can be resolved currently in v1 as well, ill drop a ref image below
![image](https://user-images.githubusercontent.com/41286589/126026994-15689402-ad8d-46f5-a87a-20ee80097bfb.png)

**Comment by JudiPuak** [comment]
Hi all:

Please excuse my ignorance.

When I run my app on Android, deep linking to calendar, phone dialer, and web browser works fine.
However, when I run it on web, none of the above works.
Is this the correct behavior or is my app missing something?
I use SDK39.

Perhaps, I need to use SDK 42 to make it work?
Or perhaps it's a future feature?

Thank you.

Judi

**Comment by mogarick** [comment]
> Hi all:
> 
> Please excuse my ignorance.
> 
> When I run my app on Android, deep linking to calendar, phone dialer, and web browser works fine.
> However, when I run it on web, none of the above works.
> Is this the correct behavior or is my app missing something?
> I use SDK39.
> 
> Perhaps, I need to use SDK 42 to make it work?
> Or perhaps it's a future feature?
> 
> Thank you.
> 
> Judi

@JudiPuak remember that v 39 is gonna be deprecated in the next release. It would be a good idea to create a new branch in your project and test v42.

> Our next release is planned for September/October 2021 and, at that time, we’ll be dropping support for SDK 39. If your project is running on SDK 39, consider upgrading to a newer version in the coming months.

Here is the link: https://blog.expo.dev/expo-sdk-42-579aee2348b6

**Comment by cronwel** [comment]
Hello Expo Team:
**Issue:** Expo Web broken, does not emulate
**Sequence:**
1. ran `npm start`
2. received prompt to install update for expo `npm install -g expo-cli`
3. reran `npm start` after installing new expo-cli`
4. pressed `w` for web interface
5. received error message below:

Failed to complie.
/Users/RBI/13-RFNK/node_modules/@react-navigation/native/lib/module/Themed.js
Module not found: Can't resolve '@react-navigation/core' in '/Users/RBI/13-RFNK/node_modules/@react-navigation/native/lib/module'

Comments:
I'm looking right at the package, following the path.
Not familiar with this type of issue, please let me know if there is something I can do to get it working again. Thanks for all you do!

**Comment by EvanBacon** [comment]
@cronwel please open an issue on https://github.com/react-navigation/react-navigation instead.

**Comment by EvanBacon** [comment]
We'll be upgrading to `react-native-web@~0.17.1` for Expo SDK 43! Here are the base differences in case you want to update your app accordingly: https://github.com/necolas/react-native-web/compare/0.13.17...0.17.1

**Comment by Maker-Mark** [comment]
@EvanBacon Is there a predictable way to get the URL history even though page refresh, or is async storage the best way to go?

**Comment by EvanBacon** [comment]
@Maker-Mark that question seems better suited for stack overflow, this issue is simply for progress tracking.

**Comment by EvanBacon** [comment]
I'm marking this issue as complete since all of the V1 tasks are done. Expo SDK components, and APIs are stable on web.

The tooling in Expo CLI (expo/webpack-config) should still be considered beta as it's not cross-platform and doesn't share enough with iOS and Android -- ideally we'd have some kinda Webpack for iOS and Android, or metro for web. This has been held up by native platform restrictions that I've been working on for a while like code splitting on native.

If you use Expo SDKs with another Webpack config like the Next.js Webpack config, then it should be stable enough for production.

I'll continue working on the bundler over in the [expo/expo-cli](https://github.com/expo/expo-cli) repo (things like [Webpack 5 upgrade](https://github.com/expo/expo-cli/pull/3763)), and [update the CLI warning](https://github.com/expo/expo-cli/pull/3766) to more specifically point to Webpack support as being the feature that is still in beta.
